ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs

```diff
@@ -85,6 +89,36 @@ data TraceLocalRootPeers peerAddr exception =
      | TraceLocalRootError   DomainAccessPoint SomeException
   deriving Show
 
+
+-- | Maximal concurrency when resolving DNS names of root and ledger peers.
+--
+maxDNSConcurrency :: Integer
+maxDNSConcurrency = 8
+
+-- | Maximal concurrency when resolving DNS names of local root peers.
+--
+maxDNSLocalRootConcurrency :: Integer
+maxDNSLocalRootConcurrency = 2
+
+-- | A semaphore used to limit concurrency of dns names resolution.
+--
+newtype DNSSemaphore m = DNSSemaphore (TSem m)
+
+-- | Create a `DNSSemaphore` for root and ledger peers.
+--
+newDNSSemaphore :: MonadSTM m => m (DNSSemaphore m)
+newDNSSemaphore = DNSSemaphore <$> atomically (newTSem maxDNSConcurrency)
```

Comment:
Maybe be explicit in the name as well: `newDNSPublicRootSemaphore`

Reply:
It's used for public roots and ledger peers, so Irenamed it to `newPublicAndLedgerRootDNSSemaphore` (pretty long name though).